### PR TITLE
Laboratory work #3, Ulyana Nagornaya - 22FPL2 

### PR DESCRIPTION
**Pull request by Ulyana-Nagornaya**

### CODE REVIEW

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def __init__(self, end_of_word_token: str) -> None:
         Args:
             end_of_word_token (str): A token denoting word boundary
         """
+        self._end_of_word_token = end_of_word_token
+        self._storage = {'_': 0}
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:27]: what is `'_'`?

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -208,6 +302,17 @@ def build(self) -> int:
...
         1 is returned
         """
+        if not isinstance(self._encoded_corpus, tuple) or len(self._encoded_corpus) == 0:
+            return 1
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+
+        if not isinstance(n_grams, tuple) or len(n_grams) == 0:
+            return 1
+
+        for n_gram in n_grams:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:312]: optimize using `set`

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -221,6 +326,20 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(sequence, tuple) or len(sequence) == 0 or len(sequence) == 0:
+            return None
+
+        freq_of_prediction = {}
+        context = sequence[self._n_gram_size - 2:]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:333]: what does the magic number meen?

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -221,6 +326,20 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
...
+
+        if len(sequence) < len(context):
+            return None
+
+        for n_gram in self._n_gram_frequencies:
+            if n_gram[:self._n_gram_size - 2] == context:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:339]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good. But I hope you will do for 10

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py:333]: resolved

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -76,6 +101,13 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or not element_id in self._storage.values():
+            return None
+
+        for index, element in enumerate(self._storage):
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:107]: use filter

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -159,8 +236,22 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or len(decoded_corpus) == 0:
+            return None
+
+        decoded_list = []
+        decoded_list += decoded_corpus[0].upper()
+
+        for element in decoded_corpus[1:-1]:
+            if element == self._end_of_word_token:
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:246]: You can do it by `replace` method

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -208,6 +302,17 @@ def build(self) -> int:
...
         """
+        if not isinstance(self._encoded_corpus, tuple) or len(self._encoded_corpus) == 0:
+            return 1
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+
+        if not isinstance(n_grams, tuple) or len(n_grams) == 0:
+            return 1
+
+        for n_gram in n_grams:
+            same_n_grams = [same for same in n_grams if same[:-1] == n_gram[:-1]]
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:313]: count?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py]: of course!

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py]: ok

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py:244]: ok

File: lab_3_generate_by_ngrams/main.py (hunk 5 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py:316]: either i don't understand or i can't imagine why we need apply this with adding extra strings(

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -104,6 +147,12 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+
+        if element not in self._storage:
+            self._storage[element] = len(self._storage)
+        return None
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:155]: There is a better way to resolve problems with mypy/pylint

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -159,7 +239,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+            return None
+
+        decoded_list = [token for token in decoded_corpus]
 
+        if decoded_list[-1] == self._end_of_word_token:
+            del decoded_list[-1]
+
+        decoded_text = ''.join(decoded_list)
+        decoded_text = decoded_text.replace(self._end_of_word_token, " ")
+        return decoded_text.capitalize() + "."
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:252]: fstring

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (artyomtugaryov) [reply lab_3_generate_by_ngrams/main.py:316]: I think, you should replace this with count

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -450,13 +719,14 @@ def get_text_processor(self) -> TextProcessor:  # type: ignore[empty-body]
         """
 
 
+
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:722]: revert

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -500,3 +770,58 @@ def _get_next_token(self, sequence_to_continue: tuple[int, ...]) -> Optional[dic
 
         In case of corrupt input arguments return None.
         """
+    """
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:773]: ? Why did you add it?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py (hunk 9 above)
Comment (Ulyana-Nagornaya) [reply lab_3_generate_by_ngrams/main.py:252]: sure
